### PR TITLE
HDS-603 Add HTML structure and preliminary Tabs styles

### PR DESCRIPTION
**Pull request by KristinLBradley**
### :pushpin: Summary

Basic preliminary HTML structure and styling for Tabs.

https://hds-components-git-hds-603-tabs-component-css-hashicorp.vercel.app/components/tabs#section2

**Notes:** 
* Styles & DOM for indicator will probably be restructured once I implement functionality.
* This branch will be merged into the [feature branch](https://github.com/hashicorp/design-system/pull/548/files) once approved. (It will not be merged directly into the `main` branch.)

### :hammer_and_wrench: Detailed description

Update: I disabled the warnings based on Melanie's feedback.

I'm getting linting errors related to accessibility that seem to me like they shouldn't be problems but would be happy for feedback.

```
8:6  error  Use of presentation role on <li> detected. Semantic elements should not be used for presentation.  no-invalid-role
20:6  error  Use of presentation role on <li> detected. Semantic elements should not be used for presentation.  no-invalid-role
10:10  error  You have an element with the role of "tab" but it is missing the required (immediate) parent element of "[tablist]". Reference: https://www.w3.org/TR/wai-aria-1.1/#tab.  require-context-role
22:10  error  You have an element with the role of "tab" but it is missing the required (immediate) parent element of "[tablist]". Reference: https://www.w3.org/TR/wai-aria-1.1/#tab.  require-context-role
```

### :camera_flash: Screenshots
Screen Shot 2022-08-30 at 2 47 46 PM

### :link: External links

***

### 👀 How to review

👉 Review commit-by-commit
👉 Review by files changed

Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

### CODE REVIEW

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [inline packages/components/addon/components/hds/tabs/index.hbs:3]: These will be split into separate child components next.

File: packages/components/app/styles/components/tabs.scss
Comment (KristinLBradley) [inline packages/components/app/styles/components/tabs.scss:32]: This may not be a pseudo-element in the final implementation when functionality is added. Current structure is just for styling purposes.

:hover is just for temporary styling purposes. Will not display on hover in final implementation.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [inline packages/components/addon/components/hds/tabs/index.hbs:6]: I used an unordered list to mark up the tab links since that's my preferred way to organize the tabs and it's also what's shown in this article that was linked from the CRD: https://inclusive-components.design/tabbed-interfaces/

The linter is complaining though. (I would prefer to keep the structure as is but of course will change it if necessary for accessibility.)

**Review by didoo — COMMENTED** [review]
@KristinLBradley added a lot of comments, let me know if you want to reply asynchronously or prefer to discuss them in a dedicated meeting.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (MelSumner) [reply packages/components/addon/components/hds/tabs/index.hbs:6]: @KristinLBradley let's turn off those two linting rules specifically for this component only.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (MelSumner) [inline packages/components/addon/components/hds/tabs/index.hbs:5]: [question] What does this wrapper do?

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (MelSumner) [inline packages/components/addon/components/hds/tabs/index.hbs:6]:
```suggestion
    <ul class="hds-tabs__tablist" role="tablist">
```

[suggestion] I think the class naming is slightly off. It should at least be `hds-tabs` first since that's the name of the component.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (MelSumner) [inline packages/components/addon/components/hds/tabs/index.hbs:8]:
```suggestion
      <li class="hds-tab" role="presentation">
```

[suggestion] since this has the role of presentation, I don't think it should get a class name. We can target it by the role.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (MelSumner) [inline packages/components/addon/components/hds/tabs/index.hbs:9]:
```suggestion
        <a
          class="hds-tabs__tab"
```

[suggestion] Since this doesn't have a presentational role, but a functional role, it should have a class name. @didoo might have stronger opinions on what it is, `hds-tabs__tab` or `hds-tabs__tablist__tab`, not 100% sure but there will be patterns in the other components to follow.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (MelSumner) [reply packages/components/addon/components/hds/tabs/index.hbs:21]: we're eliminating the use of tabs for navigation, so the markup that @KristinLBradley has so far is correct from that perspective. 👍

File: packages/components/app/styles/components/tabs.scss
Comment (heatherlarsen) [reply packages/components/app/styles/components/tabs.scss:47]: The intention is that the active state and the focus state have the same visual treatment for this component.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:5]: This is going to be separated out as a child component so I didn't want to use BEM naming style. I would have called it "hds-tabs-list" with an "s" but I was trying to follow naming in @heatherlarsen's CRD: https://docs.google.com/document/d/10OIZG32JaKxAf4uyuDjTU7SGGnwHqabBhPvfHQ8_ruA/edit#heading=h.gjdgxs

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:6]: @MelSumner  Thanks. How/where do I turn them off?

Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:6]: I'm going to break this out into separate child components as my next step. I was following naming from @heatherlarsen's CRD: https://docs.google.com/document/d/10OIZG32JaKxAf4uyuDjTU7SGGnwHqabBhPvfHQ8_ruA/edit#bookmark=id.8u2cnqhzum2o

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:8]: I'm going to break these out into child components. I was trying to follow naming in the CRD: https://docs.google.com/document/d/10OIZG32JaKxAf4uyuDjTU7SGGnwHqabBhPvfHQ8_ruA/edit#bookmark=id.8u2cnqhzum2o

Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:8]: Ok thanks but I'm going to break this out into a child component. I'll see, but I think I would want to keep a class name for the component's root element.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:9]: Thanks, I'm going to break these out into separate child components next so that may affect the naming.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:17]: Sorry I was testing and forgot to remove. Will remove

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:20]: Thanks @MelSumner

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:35]: I was following naming of the CRD and I also intend to break this out into a child component so didn't think those should use BEM naming

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:69]: Ok I think of that as a boolean.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:71]: I don't know yet if that needs to be a separate component. I'll break it out later when/if needed. I was trying to capture the main structure

File: packages/components/app/styles/components/tabs.scss
Comment (KristinLBradley) [reply packages/components/app/styles/components/tabs.scss:13]: Ok I'll add those and refactor if needed when I break into child components

File: packages/components/app/styles/components/tabs.scss
Comment (KristinLBradley) [reply packages/components/app/styles/components/tabs.scss:14]: ok I'll round it, thanks.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (heatherlarsen) [reply packages/components/addon/components/hds/tabs/index.hbs:5]: @KristinLBradley to clarify this is not "my" CRD. The intention is that the CRDs are a collaboration between Designer and Engineer. One individual starts it (doesn't matter who), the other makes changes or reviews it and comments (whatever your preferred style is), and we discuss. 🙂 

I used `tablist` in the RFC because it's the semantic name for the role.

File: packages/components/app/styles/components/tabs.scss
Comment (KristinLBradley) [reply packages/components/app/styles/components/tabs.scss:32]: Changed it to show on :first-cild. Again temporary just for working on visual styling

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (didoo) [reply packages/components/addon/components/hds/tabs/index.hbs:71]: > I don't know yet if that needs to be a separate component
Unless I'm missing something, there's no other way: it's shared between tabs, and its `left/right` position values are set via JS (so it can't be a pseudo element; unless we use local CSS props, but in that case I don't know if the animation would work... 🤔)

File: packages/components/app/styles/components/tabs.scss
Comment (KristinLBradley) [reply packages/components/app/styles/components/tabs.scss:36]: Visually it's exactly the same. I'm following the Figma in this case but I would normally set something like this to a much higher pixel value so that it would still have the same rounded appearance even if the height were to change.

File: packages/components/app/styles/components/tabs.scss
Comment (KristinLBradley) [reply packages/components/app/styles/components/tabs.scss]: I was positioning it to the bottom 1st and then using margin to pull it below. When I was initially styling it I was doing something differently that wasn't working by just using "bottom" on itself for some reason. I updated it now.

File: packages/components/app/styles/components/tabs.scss
Comment (KristinLBradley) [reply packages/components/app/styles/components/tabs.scss:57]: Ok thanks I will try it. The focus ring was getting cut off due to the overflow style on the parent so that's why I was adding the padding like that

Comment (didoo) [reply packages/components/app/styles/components/tabs.scss:57]: > The focus ring was getting cut off due to the overflow style on the parent
Using the mixin, you can set the `top/right/bottom/left` position values, so this should not be a problem.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:5]: Thanks, the naming in the CRD makes sense so I tried to follow it.

Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:5]: It does nothing currently but eventually I'll need to add overflow navigation so it would be a wrapper to contain that navigation together with the tabs. Some parts of the DOM may get slightly refactored as I work to implement all the various parts though...

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:20]: Thanks @didoo, I'm going to keep the structure as is for now. I don't think this level of navigation necessarily requires a nav element (it's mainly used for primary page navigation) and I don't want to misuse li's to trigger behavior instead of using semantic elements such as an anchor tag or button.

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:71]: Ok fine but this was meant to be a preliminary break down of the main components and it was just a comment I included to think through things as I worked. I'm going to delete the comment eventually

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:69]: Specifically, I mean I think of this behavior as a "boolean" and this comment was just for me to begin to think and plan things out:  "if they're not undefined, then you know they should be rendered" IE your comment. is how I intended to implement it, I wouldn't require someone to literally set a value of false when using the component.

The comment I wrote to plan things isn't meant to represent literally how I would implement everything in JavaScript or wherever. Once I get to the JavaScript portion then please comment on how I implement that portion if you have feedback but there's no need to pick apart a rough comment I wrote at that level...

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:73]: I updated my comment to clarify it's for temporary rough planning purposes only and not to meant to literally be the final implementation

File: packages/components/app/styles/components/tabs.scss
Comment (KristinLBradley) [reply packages/components/app/styles/components/tabs.scss:16]: Some of this may be moved and restructured as I break this out into separate components and when I add the overflow navigation since I wasn't going to work on that part until after I get the basics working

File: packages/components/app/styles/components/tabs.scss
Comment (KristinLBradley) [reply packages/components/app/styles/components/tabs.scss]: Some of this may be moved and restructured as I separate out child components

File: packages/components/app/styles/components/tabs.scss
Comment (KristinLBradley) [reply packages/components/app/styles/components/tabs.scss]: Thanks I refactored it

File: packages/components/app/styles/components/tabs.scss
Comment (KristinLBradley) [reply packages/components/app/styles/components/tabs.scss:45]: @heatherlarsen so when a user clicks on a tab (makes it active) it should show the focus ring?

File: packages/components/addon/components/hds/tabs/index.hbs
Comment (KristinLBradley) [reply packages/components/addon/components/hds/tabs/index.hbs:9]: Sorry I figured out how to turn them off. Thanks.

**Comment by KristinLBradley** [comment]
The class names are still preliminary as I was trying to plan for sub-components and also trying to follow the naming in the CRD. As I break things out into sub-components I'll look more closely and how structures have been done for other components with child-components and refactor as makes sense then...

**Comment by didoo** [comment]
> The class names are still preliminary as I was trying to plan for sub-components and also trying to follow the naming in the CRD. As I break things out into sub-components I'll look more closely and how structures have been done for other components with child-components and refactor as makes sense then...

@KristinLBradley I saw your comments/replies, but at this point I am a bit confused on 1) what to review and 2) when to approve the PR. If everything is preliminary and subject to change, why need for a review? and in that case, why need for an approval? In any case let's discuss in person, is better.

**Comment by KristinLBradley** [comment]
> 



> > The class names are still preliminary as I was trying to plan for sub-components and also trying to follow the naming in the CRD. As I break things out into sub-components I'll look more closely and how structures have been done for other components with child-components and refactor as makes sense then...
> 
> @KristinLBradley I saw your comments/replies, but at this point I am a bit confused on 1) what to review and 2) when to approve the PR. If everything is preliminary and subject to change, why need for a review? and in that case, why need for an approval? In any case let's discuss in person, is better.

@Dhaulagiri suggested I break PRs for Tabs down into a main feature branch with PRs off of that branch which would be merged back into it since smaller PRs would be easier for people to review. I'm just trying to follow that. I don't know how I can get the code in every PR to be 100% final before they are merged to the feature branch. I think if I can just get visibility on what I'm doing and feedback on my overall approach and on any specific issues that should be ok. I'm just trying to adapt to how the team generally works still.

**Review by didoo — APPROVED** [review]
As discussed in person with @KristinLBradley this is OK for now to merge. Kristin will branch again from the feature branch, and open a new PR with specific points for which she wants review, and we will continue this way for the next iterations, and then will decide how to proceed for more beefy parts of the implementation.

**Pull request merged**
